server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -2923,6 +2944,87 @@ public int deleteUpgradeSegmentsForTask(final String taskId)
     );
   }
 
+  @Override
+  public Map<String, String> retrieveUpgradedFromSegmentIds(
+      final String dataSource,
+      final Set<String> segmentIds
+  )
+  {
+    if (segmentIds.isEmpty()) {
+      return Collections.emptyMap();
+    }
+
+    final List<String> segmentIdList = ImmutableList.copyOf(segmentIds);
+    final String sql = StringUtils.format(
+        "SELECT id, upgraded_from_segment_id FROM %s WHERE dataSource = :dataSource %s",
+        dbTables.getSegmentsTable(),
+        SqlSegmentsMetadataQuery.getParameterizedInConditionForColumn("id", segmentIdList)
+    );
+    final Map<String, String> upgradedFromSegmentIds = new HashMap<>();
+    connector.retryWithHandle(
+        handle -> {
+          Query<Map<String, Object>> query = handle.createQuery(sql)
```

Comment:
Follow up work:
This query should probably be executed on batches of 100 segment IDs at a time.

Reply:
Isn't this controlled by the batch size of kill tasks to some extent?

Reply:
Yes, it would be controlled by the batch size of kill. But it is still possible for someone to either increase those limits or just fire this action with a large set of segment IDs. The overlord side should have its own safeguards.